rust/agents/validator/src/submit.rs

```diff
@@ -83,8 +88,66 @@ impl ValidatorSubmitter {
                 latest_checkpoint
             };
 
-            // ingest available messages from DB
-            while let Some(insertion) = self
+            // This can only happen if target_checkpoint is None and the latest checkpoint
+            // fetched from onchain has an index that is behind the tree's index.
+            // This may occur e.g. if RPC providers are unreliable and make calls against
+            // inconsistent block tips.
+            //
+            // In this case, we just sleep a bit until we fetch a new latest checkpoint
+            // that at least meets the tree.
+            if correctness_checkpoint.index < tree.index() {
+                sleep(self.interval).await;
+                break;
+            }
+
+            self.submit_checkpoints_until_correctness_checkpoint(
+                &mut tree,
+                &correctness_checkpoint,
+            )
+            .await?;
+
+            // If target checkpoint is Some and we've gotten this far,
+            // we've successfully reached the target checkpoint and break.
+            if target_checkpoint.is_some() {
+                info!(?target_checkpoint, "Successfully reached target checkpoint");
+                break;
+            } else {
+                // Only update the latest checkpoint processed metric if we're
+                // not targeting a specific checkpoint and instead want to keep
+                // signing checkpoints indefinitely as they come in.
+                self.metrics
+                    .latest_checkpoint_processed
+                    .set(correctness_checkpoint.index as i64);
+            }
+
+            sleep(self.interval).await;
+        }
+
+        // TODO: remove this once validator is tolerant of tasks exiting.
+        loop {
+            sleep(Duration::from_secs(u64::MAX)).await;
+        }
+    }
+
+    /// Submits signed checkpoints relating to the given tree until the correctness checkpoint (inclusive).
+    /// Only submits the signed checkpoints once the correctness checkpoint is reached.
+    async fn submit_checkpoints_until_correctness_checkpoint(
+        &self,
+        tree: &mut IncrementalMerkle,
+        correctness_checkpoint: &Checkpoint,
+    ) -> Result<()> {
+        // This should never be called with a tree that is ahead of the correctness checkpoint.
+        assert!(correctness_checkpoint.index >= tree.index());
+
+        // All intermediate checkpoints will be stored here and signed once the correctness
+        // checkpoint is reached.
+        let mut checkpoint_queue = vec![];
+
+        // If the correctness checkpoint is ahead of the tree, we need to ingest more messages.
+        //
```

Comment:
Nit: >= is more intuitive to me

Reply:
Though in the spirit of this PR, I feel like if strictly greater than occurs, then that's maybe worth logging

Reply:
Personally feel that converting the index to a count with the + 1 makes it easier to understand, I'll make a better comment too